### PR TITLE
Wring out extension parsing

### PR DESCRIPTION
**Pull request by CyberiaResurrection**
With TradeCodes, StarList, Nobles and UWP all hoisted out to their own objects and having their parsing wrung out, that's leaving extension parsing looking a little ratty.

This PR doesn't hook up an Extension object, but it does shine a light onto the darker corners of Importance/Economic/Cultural extension parsing and not leave _quite_ so much to chance.

### CODE REVIEW

**Comment by CyberiaResurrection** [comment]
> The `_find_importance` procedure started to set off some red flags. In reviewing the test data I figured out why. The assumption in the original `starline` regex is that the line in the data will contain all three of the Ix, Ex, and Cx, or will not have any of them and they will be represented by blank spaces.

How does that play with the existing code _generating_ Ix, come what may?
My concern is when going round-trip, frinstance:
`2720 Vlazzhden            C210143-8 Lo Ni                               - -  - 303   Zh G1 IV                       `
gets parsed and returned as
`2720 Vlazzhden            C210143-8 Lo Ni                                 { -2 } -       -      -    -  - 303 0  Zh   G1 IV                                                 `


> 
> Your code says that assumption is wrong, and the test cases are there to validate this.

I was generating the candidate lines off the starline regex, so such a line would make it past the regex and have to be parsed.
> 
> My response to that is either:
> 
> 1. The three data (Ix, Ex, Cx) are part of the T5 rules and mandated by the T5 Second Survey. If a sector is in review or approved all three of these will be there, and if unreviewed they may not be. But there is no sector where there will be one or two but not all three.
> 2. We should follow Postel's law: "be conservative in what you send, be liberal in what you accept", and fix the starline regex. Specifically line 20 should be split into three parts (one for each of the Ix, Ex, and Cx). Unfortunately, due to the way the groups are written the following group numbers may be off if you fix this

1 - If extensions are all or nothing (which you're saying they are), where does the existing `calculate_importance()` call fit in?  Does Ix only get spat out if both Ex and Cx are present?  On the way in, should Ex and Cx only be filled if _both_ are present?
2 - My regex-fu is fairly weak, so I'm leery of mucking with the starline regex.

**Comment by tjoneslo** [comment]
> 1 - If extensions are all or nothing (which you're saying they are), where does the existing calculate_importance() call fit in? Does Ix only get spat out if both Ex and Cx are present? On the way in, should Ex and Cx only be filled if both are present?

In my remembering of the code prior to some of your changes, the `calculate_importance()` was only called when the source had an Ix to begin with. This was true of the Ex and Cx as well. It might be, and I'd have to double check, the code would calculate the Importance regardless because 1. it was static based upon other data (no random numbers) and 2. ended up being used for the Eaglestone Trade Index, and generating routes for that. 

> 2 - My regex-fu is fairly weak, so I'm leery of mucking with the starline regex.
Largely this is splitting line 20 into three parts. Let me find some time with the regex parser tools and I'll send you an updated version.

**Comment by CyberiaResurrection** [comment]
> > 1 - If extensions are all or nothing (which you're saying they are), where does the existing calculate_importance() call fit in? Does Ix only get spat out if both Ex and Cx are present? On the way in, should Ex and Cx only be filled if both are present?
> 
> In my remembering of the code prior to some of your changes, the `calculate_importance()` was only called when the source had an Ix to begin with. This was true of the Ex and Cx as well. It might be, and I'd have to double check, the code would calculate the Importance regardless because 1. it was static based upon other data (no random numbers) and 2. ended up being used for the Eaglestone Trade Index, and generating routes for that.

Star parsing has unconditionally generated Ix since at least 
https://github.com/makhidkarun/traveller_pyroute/blob/d78285f5029f5bbeb1bd39e485be052c95e336eb/PyRoute/Star.py

That commit landed no later than Jul 9, 2019 - predating the _vast_ bulk of the changes I've landed (except speeding up Star __hash__ and __eq__, I think).

> 
> > 2 - My regex-fu is fairly weak, so I'm leery of mucking with the starline regex.
> > Largely this is splitting line 20 into three parts. Let me find some time with the regex parser tools and I'll send you an updated version.

Were you thinking about the updated copy you provided upthread?

**Comment by tjoneslo** [comment]
I pulled up my favorite [regex testing](https://regex101.com/) site and put in this (slightly altered) version of line 20, and asked it to match several of the versions possible Ix, Ex, Cx 

```((\{ *[+-]?[0-6] ?\}|-) +(\([0-9A-Z]{3}[+-]\d\)|-) +(\[[0-9A-Z]{4}\]|-)|( ) ( ) ( )) +```

The regex, according to the site, correctly matched all of these, with the values either being a `-` or the values in the correct delimiters. The final set `( ) ( ) ( )` is for the case of the non-T5 sectors where the Traveller Map generates spaces for each of these values when they are missing entirely. 
```
{0} (000-0) [0000] 
{0} - [0000] 
{0} - - 
{0} (000-0) - 
- - - 
- (000-0) - 
- - [0000] 
- (000-0) [0000] 
```
So either the Ix, Ex, Cx are generated with values or "-" for missing values. Or all three are missing entirely and the system will match 3 spaces, separated by 3 more spaces (total of 6 or more spaces) consumed from the line.

**Comment by CyberiaResurrection** [comment]
Maybe an example would make it easier.
`2618 Horden 2618          D54A367-D Ht Lo Wa O:2915                     - - - 401   So       D `
note complete lack of extension gubbins -
parses to line as:
`2618 Horden 2618          D54A367-D Ht Lo O:2915 Wa                       { -1 } -       -      -    -  - 401 0  So   D                                                       `

Is this what it _should_ be doing when the source has no extension gubbins?

**Comment by tjoneslo** [comment]
> Is this what it should be doing (inserting calculated importance, and dash Ex/Cx) when the source has no extension gubbins?

This is a really good question. The original in the code assumed three cases: 
1. Older sectors with no T5 review would have only spaces in the Ix, Ex, Cx positions (the second half of the regex). 
2. Full review sectors with all three T5 extensions, for which we can parse out the values
3. A sector under review which would have the Ix, but not the Ex and Cx values. They would be `-`. 

The final case of having no Ix, Ex, Cx values with `-` in all three positions was never considered. You would either have a star line with an Ix and optional Ex and Cx values as `-`. Or all three were spaces.  

```
((\{ *[+-]?[0-6] ?\}) +(\([0-9A-Z]{3}[+-]\d\)|-) +(\[[0-9A-Z]{4}\]|-)|( ) ( ) ( )) +
```
This is regex that does that. Note: Slightly changed from above.

So what are the other edge cases we need to capture?

**Comment by CyberiaResurrection** [comment]
> > Is this what it should be doing (inserting calculated importance, and dash Ex/Cx) when the source has no extension gubbins?
> 
> This is a really good question. The original in the code assumed three cases:
> 
> 1. Older sectors with no T5 review would have only spaces in the Ix, Ex, Cx positions (the second half of the regex).
> 2. Full review sectors with all three T5 extensions, for which we can parse out the values
> 3. A sector under review which would have the Ix, but not the Ex and Cx values. They would be `-`.
> 
> The final case of having no Ix, Ex, Cx values with `-` in all three positions was never considered. You would either have a star line with an Ix and optional Ex and Cx values as `-`. Or all three were spaces.
> 
> ```
> ((\{ *[+-]?[0-6] ?\}) +(\([0-9A-Z]{3}[+-]\d\)|-) +(\[[0-9A-Z]{4}\]|-)|( ) ( ) ( )) +
> ```
> 
> This is regex that does that. Note: Slightly changed from above.
> 
> So what are the other edge cases we need to capture?

I'm not sure if the current setup's output can even _distinguish_ between all-spaces and all-dashes extension values - they get spat out as dashes if none, their value otherwise.

Your case 2 is a walk-up start, having been handled for yonks.
Cases 1 and 3 are the ones bothering me.

My initial idea was to blank/dash out Ix when parsing to line if foo.economics and foo.social are both None.  That _sort of_ handles case 1, and points and laughs at case 3.

Adding (another) boolean field to Star, `_oldskool` is about the only way I can see of easily handling this.
Iff all three importance fields are spaces (not none, not dash, etc), then set `_oldskool` to True.

On the way out (`parse_to_line()`), `_oldskool` being True replaces all three importance fields with spaces, while False means continue with the status quo, where Ix is unconditionally output, with Ex/Cx being dashed if empty, their values otherwise.

**Comment by CyberiaResurrection** [comment]
I've implemented `_oldskool` field and its matching property.

As a result, `2618 Horden 2618          D54A367-D Ht Lo Wa O:2915                     - - - 401   So       D ` is now handled as an oldskool line - its parse_to_line output also lacks all three of Ix, Cx and Ex.

How should the following be handled?
`2618 Horden 2618          D54A367-D Ht Lo Wa O:2915       - - -         - - - 401   So       D `   - each extension being dashed instead of a space.  That new line would _seem_ semantically equivalent to the original - no Ix, Cx and/or Ex - but should it be counted as oldskool?

**Comment by tjoneslo** [comment]
> How should the following be handled?
> `2618 Horden 2618 D54A367-D Ht Lo Wa O:2915 - - - - - - 401 So D ` - each extension being dashed instead of a space. That new line would _seem_ semantically equivalent to the original - no Ix, Cx and/or Ex, but in a different form - but should it be counted as oldskool?

The current regex will fail trying to parse this line. And I'm good with that. You should never output a line like this. 

On output: If you have a full UWP, and can generate an Ix score, output the Ix and two dashes. If the UWP has any `?`s, and not all the data is known even if it's not used by the Ix score generation, assume old school and output spaces. 

I'm also coming back to the idea of not having additional parsing of the star line. That is, the regex always correctly parses the star line into it's components. If it can't it means either the regex is wrong and should be fixed, or the data is bad and should be rejected. Since the regex wants to reject the line above, its bad data and needs to be fixed by a human.

**Comment by CyberiaResurrection** [comment]
I've remixed the oldskool property as you asked - it's now driven by an eponymous property on the UWP object.

With the auxiliary parsing commented out, I can give you the following starlines (matching the starline regex with your new line 20) that have/cause problems with extension parsing at the moment (not line output), not getting the extension(s) they should seem to get:

```
0101 000000000000000 ???????-? 000000000000000 {0} -  [0000]         - 0 000   0000D
0101 000000000000000 ???????-? 000000000000000 {0} -  [0000] -         - 0 000   0000D
0101 000000000000000 ???????-? 000000000000000 {0 } (000-0) -         - 0 000   0000D
0101 000000000000000 ???????-? 000000000000000 {0} -  [0000] -         - 0 000   0000D
0101 000000000000000 ???????-? 000000000000000 { 0} -  [0000]         - 0 000   0000D
0101 000000000000000 ???????-? 000000000000000 { 0 } -  [0000]         - 0 000   0000D
0101 000000000000000 ???????-? 000000000000 {0 {0} -  [0000] -         - 0 000   00+D
0101 000000000000000 ???????-? 0000000000000 { {0} -  [0000] -         - 0 000   00+D
0101 000000000000000 ???????-? 000000000000000 { 0 } -  [000a]         - 0 000   0000D
0101 000000000000000 ???????-? 000000000000000 {0 } (a00-0) -         - 0 000   0000D
0101 000000000000000 ???????-? 000000000000000 {0} (000-0)       -    - 0 000   0000A0
```

**Comment by bobbyjim** [comment]
I've been working on a Rakulang grammar that parses the UWP line properly,
with grammar rules.  This overcomes the painful limitations of regexen.

**Comment by CyberiaResurrection** [comment]
> I've been working on a Rakulang grammar that parses the UWP line properly, with grammar rules. This overcomes the painful limitations of regexen.

Standing alone, that sounds pretty awesome, @bobbyjim .  How would your proposed UWP parser bolt into the rest of PyRoute?

**Comment by CyberiaResurrection** [comment]
> The `_find_importance` procedure started to set off some red flags. In reviewing the test data I figured out why. The assumption in the original `starline` regex is that the line in the data will contain all three of the Ix, Ex, and Cx, or will not have any of them and they will be represented by blank spaces.
> 

That `_find_importance` method is now gone, so does this individual review hold weight any more?

**Comment by bobbyjim** [comment]
Need help?  I'm a regex fanatic.

**Review by tjoneslo — APPROVED** [review]

File: PyRoute/Inputs/StarlineParser.py [hunk 1]
```diff
@@ -31,11 +31,11 @@ class StarlineParser:
     MAJOR_SOPHONT.3: /\[[^\]\{]{1,}\][W\d\?]{0,1}/
     OWNED_COLONY.3: /[OC]:[X\d\?]{0,4}/ | /[OC]:[A-Z][A-Za-z]{3,3}[-\:]{0,1}\d{4,4}/
     RESIDUAL.2: /[0-9A-Za-z?\-+*()\'\{\}\[\]]{2,}/
-    SINGLETON: /[0-9A-Za-z\+\*()?\']/
+    SINGLETON: /[0-9AC-Za-z\+\*()?\']/
```
[the remark this answers is not in the document]
Comment (CyberiaResurrection) [reply PyRoute/Inputs/StarlineParser.py:34]: To remove collisions between single noble codes of "B" and tail end singleton trade codes.

**Pull request merged**